Sources/BrowserServicesKit/ContentBlocking/ContentBlockerRulesManager.swift

```diff
@@ -25,11 +25,109 @@ import TrackerRadarKit
 
     func rulesManager(_ manager: ContentBlockerRulesManager,
-                      didUpdateRules: ContentBlockerRulesManager.CurrentRules,
-                      changes: ContentBlockerRulesIdentifier.Difference,
+                      didUpdateRules: [ContentBlockerRulesManager.Rules],
+                      changes: [String: ContentBlockerRulesIdentifier.Difference],
                       completionTokens: [ContentBlockerRulesManager.CompletionToken])
 }
 
+/**
+ Encapsulates compilation steps for a single Task
+ */
+private class CompilationTask {
+    typealias Completion = () -> Void
+    let workQueue: DispatchQueue
+    let rulesList: ContentBlockerRulesList
+    let sourceManager: ContentBlockerRulesSourceManager
+    let logger: OSLog
+
+    var completed: Bool { result != nil }
+    var result: (compiledRulesList: WKContentRuleList, model: ContentBlockerRulesSourceModel)?
+
+    init(workQueue: DispatchQueue,
+         rulesList: ContentBlockerRulesList,
+         sourceManager: ContentBlockerRulesSourceManager,
+         logger: OSLog = .disabled) {
+        self.workQueue = workQueue
+        self.rulesList = rulesList
+        self.sourceManager = sourceManager
+        self.logger = logger
+    }
+
+    func start(completionHandler: @escaping Completion) {
+
+        let model = sourceManager.makeModel()
+
+        // Delegate querying to main thread - crashes were observed in background.
+        DispatchQueue.main.async {
+            WKContentRuleListStore.default()?.lookUpContentRuleList(forIdentifier: model.rulesIdentifier.stringValue,
+                                                                    completionHandler: { ruleList, _ in
+                if let ruleList = ruleList {
+                    self.compilationSucceded(with: ruleList, model: model, completionHandler: completionHandler)
+                } else {
+                    self.workQueue.async {
+                        self.compile(model: model, completionHandler: completionHandler)
+                    }
+                }
+            })
+        }
+    }
+    
+    private func compilationSucceded(with compiledRulesList: WKContentRuleList,
+                                     model: ContentBlockerRulesSourceModel,
+                                     completionHandler: @escaping Completion) {
+        workQueue.async {
+            self.result = (compiledRulesList, model)
+            completionHandler()
+        }
+    }
+
+    private func compilationFailed(for model: ContentBlockerRulesSourceModel,
+                                   with error: Error,
+                                   completionHandler: @escaping Completion) {
+        os_log("Failed to compile %{public}s rules %{public}s", log: logger, type: .error, rulesList.name, error.localizedDescription)
+
+        // Retry after marking failed state in the source
+        sourceManager.compilationFailed(for: model, with: error)
+        let newModel = sourceManager.makeModel()
```

Comment:
I think this should go inside self.workQueue.async { .. } as sourceManager.makeModel() usage is made from the workQueue and the compilationFailed callback may be called by the compileContentRuleList from the main thread

Reply:
It's not like there's a bug - as operations here are done sequentially and are encapsulated to a single class instance - but this is something I *know* as an author of this class :) and it isn't necessary a good thing. I agree that having this in a async block is more clear to the reader and possibly less error prone in case of any refactoring - going to change this.